src/core/connection.c

```diff
@@ -5385,8 +5389,11 @@ QuicConnRecvDatagramBatch(
         CXPLAT_DBG_ASSERT(Packet->PacketId != 0);
         if (QuicConnRecvPrepareDecrypt(
                 Connection, Packet, HpMask + i * CXPLAT_HP_SAMPLE_LENGTH) &&
-            QuicConnRecvDecryptAndAuthenticate(Connection, Path, Packet) &&
-            QuicConnRecvFrames(Connection, Path, Packet, ECN)) {
+            QuicConnRecvDecryptAndAuthenticate(Connection, Path, Packet)) {
+
+            if (!QuicConnRecvFrames(Connection, Path, Packet, ECN)) {
+                continue;
+            }
```

Comment:
I feel like there must be a better way to clean this code up. Perhaps moving the `else` code below (line 5418) somewhere else?

Reply:
The issue is that we need to rollback compatible version negotiation on decryption failure. The `QuicConnRecvFrames` failure path here convolutes the failure path.

Reply:
I've been thinking about this, and the options are having the code in two places (I don't like this because it could lead to bugs in the future), or having a temporary variable in the if-statement.  
Are you more concerned about readability or the performance of this code path?

Reply:
More for readability.

Reply:
how's this for readability?